[no squash] lib: Filter out deprecated bridges from BridgeList

jtojnar opened the pull request:
Alternative to https://github.com/RSS-Bridge/rss-bridge/pull/2885.

dvikan committed:
Improve BridgeFactory variable names

jtojnar committed:
bridges: Add deprecated annotation to deprecated bridges

Once we raise minimum PHP version to 8, we can use the #[Deprecated] attribute instead.

jtojnar committed:
lib: Filter out deprecated bridges from BridgeList

jtojnar committed:
lib/BridgeFactory: Make methods only accept valid class names

This moves the responsibility for getting a valid class name
to the users of BridgeFactory, avoiding the repeated sanitation.
Improper use can be checked statically.

dvikan commented:
Thanks for the pr. There's too much going on here. We are not introducing psalm or php8 annotations on my watch. Also using reflection to grab a value from phpdoc is insane.

jtojnar commented:
> There's too much going on here.

It is meant to be reviewed commit by commit but I split it to separate PRs if that will make it easier.

> We are not introducing psalm or php8 annotations on my watch.

The PHPDoc annotations I introduced are not PHP8 nor Psalm specific – they are widely supported by tools like PHPStan, Psalm and PHPStorm. And I find them immensely useful just for looking at the functions since they allow me to immediately see what the function takes as argument/returns.

For example, when I was trying to review #2887, I had to first check where `$submissionHTML` is defined to find out its type. That lead me to `getFASimpleHTMLDOM` method, which still did not get me this information, so I continued to `defaultLinkTo` function, which according to its type annotation returns `object`. At this point I gave up in frustration. If each function (at least in `lib/`) contained precise type annotation, I could stop this traversal after one step. Or if I used PHPStorm as an IDE or LSP, I could just hover over the variable with my mouse and would learn the type immediately.

> Also using reflection to grab a value from phpdoc is insane.

I believe that is standard way for getting PHPDoc annotations. Normally, I would use something like [phpdocumentor/reflection-docblock](https://packagist.org/packages/phpdocumentor/reflection-docblock) but since I know your position on dependencies, I decided to roll a NIH implementation :wink: 

The only other somewhat clear option I can come up with is to provide this metadata is by implementing an interface like the following:

```php
interface DeprecatedBridge {
    function getDeprecationReason(): ?string;
}
```

But I would say deprecation status is not an inherent property of a bridge but rather a status conveyed to it so metadata like docblock is better place for this information. (Though this is somewhat philosophical notion and we need not follow it strictly if it clashes with some practical concerns.)

dvikan commented:
I understand your point about typing. I am in favor of more typing in the actual php code (not docblock). And yes I know that some typing must be done in phpdoc e.g. `BridgeInterface[]`.
